### PR TITLE
[B] Escape Unicode characters when dumping YAML. Fixes BUKKIT-1466

### PR DESCRIPTION
**Pull request by Jckf**
# The issue:

Bukkit's YAML library, SnakeYaml, defaults to outputting unescaped Unicode characters. This causes Bukkit to produce YAML files with characters replaced by question marks.
# Justification for this PR:

Implementing this change would ease plugin developers' work to support languages that use "special characters" and symbols.
# PR breakdown:

From SnakeYaml's documentation for setAllowUnicode():

> Specify whether to emit non-ASCII printable Unicode characters (to support ASCII terminals). The default value is true.

Setting this to false causes the Unicode characters to be escaped, and in turn Bukkit to produce a valid and loadable YAML file.
# Testing results and materials:

Loading this YAML file http://pastie.org/8380939 before applying this PR results in the following content once saveConfig() is invoked: http://pastie.org/8380940. Once the PR is applied, no apparent changes are made to the file upon saving.
# JIRA ticket

https://bukkit.atlassian.net/browse/BUKKIT-1466

### CODE REVIEW

**Comment by Wolvereness** [comment]
If the files were written and loaded in strict UTF8, would this be a problem?

/edit: this referring to, would the PR still be applicable

**Comment by Jckf** [comment]
I've tried replacing FileWriter in FileConfiguration with a OutputStreamWriter configured for UTF-8, but that only made things worse. I couldn't make heads or tails of it.

**Comment by Wolvereness** [comment]
You'd need to read it back as UTF8 as well.

**Comment by Jckf** [comment]
The resulting file made no sense in Notepad++ or HxD, so I doubt it was valid UTF-8 at all.

**Comment by Wolvereness** [comment]
Could you try the same text in a strictly UTF8 facility, like I did in https://github.com/Wolvereness/PhysicalShop/blob/master/src/main/java/com/wolvereness/physicalshop/config/Localized.java#L131 ? NP++ may need its input encoding specified (because you would have ran it from windows), you can check that at the top or refer to this [SO](http://superuser.com/questions/292086/how-can-i-enforce-so-notepad-uses-utf-8-every-time-i-create-a-new-file).

It's also worth noting that the BOM does not work in our yaml implementation; make sure you don't have it.

**Comment by Jckf** [comment]
I am unsure what you're asking me to do. Please note that my native language is not English/American. Also, I never use byte order marks.

**Comment by Wolvereness** [comment]
TL;DR: could you take the offending text, _saved as UTF8_ (double check the setting in NP++), into PhysicalShop?

**Comment by Jckf** [comment]
Using norwegian characters in the locale files for PhysicalShop produces no stack trace or error message in the console, but the result in-game is not as expected. I see single characters replaced by 2 character sequences starting with the letter Ã.

**Comment by Wolvereness** [comment]
PhysicalShop will overwrite the file, does the new file appear as you wrote it? This may be an issue with your UTF8 editor.

**Comment by Jckf** [comment]
I see no changes in the file itself.

**Comment by Wolvereness** [comment]
https://gist.github.com/Wolvereness/6849779

/edit: To clarify, that test verifies that the current system is working as intended, albeit broken. The problem, as observed, is an issue with expecting the server to be UTF8 (which Bukkit doesn't force, for legacy reasons).

/edit2: I used base64 encoding to prevent transcoding when copy+pasting or viewing on the web (which don't retain UTF8). The two examples were ones that @Jckf provided me over IRC, zipped. The test case includes means for producing the base64 file if you can make a file (plaintext) locally.

**Comment by Jckf** [comment]
You're saying that loading escape sequences and saving the characters as question marks is intended, and that it won't be fixed for legacy reasons, which leaves Bukkit with a broken configuration system that doesn't really support any characters outside the ASCII table, even when properly escaped. Users that require localization or have other reasons to include "special characters" or symbols in their config are therefore forced to write their own configuration handling.

EDIT: This PR does not break existing behavior. It only adds support for escape sequences. So I do not see the reason to close it.

EDIT 2: Unless you expect that people are relying on their escape sequences turning into literal question marks, and want to continue supporting it =P

**Comment by Wolvereness** [comment]
I'm saying that, you gave me test files, and those test files _actually work_ when you set your editor to the correct encoding.

**Comment by Jckf** [comment]
From Windows: https://www.dropbox.com/s/2n8vnvutzdocl6t/encoding-win.zip
From Linux: https://www.dropbox.com/s/1qa5428e35eerhe/encoding-nix.zip

**Comment by Wolvereness** [comment]
The windows box of yours worked successfully.
The linux box doesn't have locale set up and uses strict ASCII by default. Here's the log:

```
file1.yml
target/encoding/default-US-ASCII
23 20 41 45 20 5C 78 43 36 0A 23 20 4F 45 20 5C 78 44 38 0A 23 20 41 41 20 5C 78 43 35 0A 23 20 61 65 20 5C 78 45 36 0A 23 20 6F 65 20 5C 78 46 38 0A 23 20 61 61 20 5C 78 45 35 0A 0A 6D 65 73 73 61 67 65 3A 0A 2D 20 27 26 36 56 65 6C 6B 6F 6D 6D 65 6E 20 74 69 6C 20 56 6F 58 2C 20 26 66 40 70 26 36 21 27 0A 2D 20 27 26 36 48 65 72 20 68 61 72 20 76 69 20 65 6E 20 6E 65 64 72 65 20 61 6C 64 65 72 73 67 72 65 6E 73 65 20 70 3F 20 26 63 31 38 20 3F 72 26 36 2E 27 0A 2D 20 27 26 36 53 70 69 6C 6C 65 72 65 20 73 6F 6D 20 65 72 2C 20 65 6C 6C 65 72 20 26 6F 6F 70 70 66 3F 72 65 72 26 72 26 36 20 73 65 67 20 73 6F 6D 20 6F 6D 20 64 65 20 65 72 20 79 6E 67 72 65 2C 20 76 69 6C 20 75 74 65 73 74 65 6E 67 65 73 2E 27 0A 2D 20 27 26 36 45 6C 6C 65 72 73 20 65 72 20 64 65 74 20 62 61 72 65 20 3F 20 67 6A 3F 72 65 20 73 6F 6D 20 6D 61 6E 20 76 69 6C 2C 20 73 3F 20 6C 65 6E 67 65 20 64 65 74 20 69 6B 6B 65 20 70 6C 61 67 65 72 20 6E 6F 65 6E 20 3A 29 27 0A 2D 20 27 27 0A 
target/encoding/default-2-US-ASCII
23 20 41 45 20 5C 78 43 36 0A 23 20 4F 45 20 5C 78 44 38 0A 23 20 41 41 20 5C 78 43 35 0A 23 20 61 65 20 5C 78 45 36 0A 23 20 6F 65 20 5C 78 46 38 0A 23 20 61 61 20 5C 78 45 35 0A 0A 6D 65 73 73 61 67 65 3A 0A 2D 20 27 26 36 56 65 6C 6B 6F 6D 6D 65 6E 20 74 69 6C 20 56 6F 58 2C 20 26 66 40 70 26 36 21 27 0A 2D 20 27 26 36 48 65 72 20 68 61 72 20 76 69 20 65 6E 20 6E 65 64 72 65 20 61 6C 64 65 72 73 67 72 65 6E 73 65 20 70 3F 20 26 63 31 38 20 3F 72 26 36 2E 27 0A 2D 20 27 26 36 53 70 69 6C 6C 65 72 65 20 73 6F 6D 20 65 72 2C 20 65 6C 6C 65 72 20 26 6F 6F 70 70 66 3F 72 65 72 26 72 26 36 20 73 65 67 20 73 6F 6D 20 6F 6D 20 64 65 20 65 72 20 79 6E 67 72 65 2C 20 76 69 6C 20 75 74 65 73 74 65 6E 67 65 73 2E 27 0A 2D 20 27 26 36 45 6C 6C 65 72 73 20 65 72 20 64 65 74 20 62 61 72 65 20 3F 20 67 6A 3F 72 65 20 73 6F 6D 20 6D 61 6E 20 76 69 6C 2C 20 73 3F 20 6C 65 6E 67 65 20 64 65 74 20 69 6B 6B 65 20 70 6C 61 67 65 72 20 6E 6F 65 6E 20 3A 29 27 0A 2D 20 27 27 0A 
target/encoding/UTF8
23 20 41 45 20 5C 78 43 36 0A 23 20 4F 45 20 5C 78 44 38 0A 23 20 41 41 20 5C 78 43 35 0A 23 20 61 65 20 5C 78 45 36 0A 23 20 6F 65 20 5C 78 46 38 0A 23 20 61 61 20 5C 78 45 35 0A 0A 6D 65 73 73 61 67 65 3A 0A 2D 20 27 26 36 56 65 6C 6B 6F 6D 6D 65 6E 20 74 69 6C 20 56 6F 58 2C 20 26 66 40 70 26 36 21 27 0A 2D 20 27 26 36 48 65 72 20 68 61 72 20 76 69 20 65 6E 20 6E 65 64 72 65 20 61 6C 64 65 72 73 67 72 65 6E 73 65 20 70 C3 A5 20 26 63 31 38 20 C3 A5 72 26 36 2E 27 0A 2D 20 27 26 36 53 70 69 6C 6C 65 72 65 20 73 6F 6D 20 65 72 2C 20 65 6C 6C 65 72 20 26 6F 6F 70 70 66 C3 B8 72 65 72 26 72 26 36 20 73 65 67 20 73 6F 6D 20 6F 6D 20 64 65 20 65 72 20 79 6E 67 72 65 2C 20 76 69 6C 20 75 74 65 73 74 65 6E 67 65 73 2E 27 0A 2D 20 27 26 36 45 6C 6C 65 72 73 20 65 72 20 64 65 74 20 62 61 72 65 20 C3 A5 20 67 6A C3 B8 72 65 20 73 6F 6D 20 6D 61 6E 20 76 69 6C 2C 20 73 C3 A5 20 6C 65 6E 67 65 20 64 65 74 20 69 6B 6B 65 20 70 6C 61 67 65 72 20 6E 6F 65 6E 20 3A 29 27 0A 2D 20 27 27 0A 
target/encoding/UTF8-2
23 20 41 45 20 5C 78 43 36 0A 23 20 4F 45 20 5C 78 44 38 0A 23 20 41 41 20 5C 78 43 35 0A 23 20 61 65 20 5C 78 45 36 0A 23 20 6F 65 20 5C 78 46 38 0A 23 20 61 61 20 5C 78 45 35 0A 0A 6D 65 73 73 61 67 65 3A 0A 2D 20 27 26 36 56 65 6C 6B 6F 6D 6D 65 6E 20 74 69 6C 20 56 6F 58 2C 20 26 66 40 70 26 36 21 27 0A 2D 20 27 26 36 48 65 72 20 68 61 72 20 76 69 20 65 6E 20 6E 65 64 72 65 20 61 6C 64 65 72 73 67 72 65 6E 73 65 20 70 3F 20 26 63 31 38 20 3F 72 26 36 2E 27 0A 2D 20 27 26 36 53 70 69 6C 6C 65 72 65 20 73 6F 6D 20 65 72 2C 20 65 6C 6C 65 72 20 26 6F 6F 70 70 66 3F 72 65 72 26 72 26 36 20 73 65 67 20 73 6F 6D 20 6F 6D 20 64 65 20 65 72 20 79 6E 67 72 65 2C 20 76 69 6C 20 75 74 65 73 74 65 6E 67 65 73 2E 27 0A 2D 20 27 26 36 45 6C 6C 65 72 73 20 65 72 20 64 65 74 20 62 61 72 65 20 3F 20 67 6A 3F 72 65 20 73 6F 6D 20 6D 61 6E 20 76 69 6C 2C 20 73 3F 20 6C 65 6E 67 65 20 64 65 74 20 69 6B 6B 65 20 70 6C 61 67 65 72 20 6E 6F 65 6E 20 3A 29 27 0A 2D 20 27 27 0A 
```

If that box had a locale set up properly, or a -Ddefault.ecoding, it wouldn't have been lossy, as evidenced by target/encoding/UTF8. The other three were the same result of the data being converted to ascii.

**Comment by Jckf** [comment]
I assume you mean -Dfile.encoding. I haven't tested the result of running with that, without this PR, yet.
